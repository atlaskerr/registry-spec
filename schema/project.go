package schema

import (
	"github.com/xeipuuv/gojsonschema"
)

func project() *gojsonschema.Schema {
	path := "project/project.schema.json"
	mediaType := MediaTypeProject
	return load(path, mediaType)
}

func projectCreate() *gojsonschema.Schema {
	path := "project/project.create.schema.json"
	mediaType := MediaTypeProjectCreate
	return load(path, mediaType)
}

func projectList() *gojsonschema.Schema {
	path := "project/project.list.schema.json"
	mediaType := MediaTypeProjectList
	return load(path, mediaType)
}
