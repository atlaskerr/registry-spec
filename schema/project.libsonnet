// Copyright Â© 2019 Titan Distribution Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

local types = import 'types.libsonnet';

local openapi = 'openapi';
local jsonschema = 'jsonschema';
local V7 = 'http://json-schema.org/draft-07/schema#';

local jid = {
  project: 'https://openpackages.io/schema/project',
  createProject: 'https://openpackages.io/schema/project-create',
  listProjects: 'https://openpackages.io/schema/project-list',
};

local projectSchemaFunc(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.project,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  properties: {
    name: { type: 'string' },
    namespace: { type: 'string' },
    repositories: types.uint64,
    labels: types.labels(output),
  },
};

local projectExample = {
  name: 'project-foo',
  namespace: 'namespace-foo',
  labels: {
    provider: {
      awsAccount: 'aws-account-foo',
      pricingPlan: 'PREMIUM',
      region: 'us-east-1',
    },
    consumer: {
      team: 'team-foo',
      manager: 'Tom Ripen',
      costCenter: 'cs-foo',
    },
  },
};

local projectCreateSchemaFunc(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.createProject,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  properties: {
    labels: types.mapStringString(output),
  },
};

local projectCreateExample = {
  labels: {
    provider: {
      awsAccount: 'aws-account-foo',
      pricingPlan: 'PREMIUM',
      region: 'us-east-1',
    },
    consumer: {
      team: 'team-foo',
      manager: 'Tom Ripen',
      costCenter: 'cs-foo',
    },
  },
};

local projectListSchemaFunc(output=jsonschema) = {
  [if output == jsonschema then '$id']: jid.listProjects,
  [if output == jsonschema then '$schema']: V7,
  type: 'object',
  properties: {
    projects: {
      type: 'array',
      items: projectSchemaFunc(''),
    },
  },
};

local projectListExample = {
  projects: [
    {
      name: 'project-foo',
      namespace: 'namespace-foo',
      repositories: 8,
      labels: {
        provider: {
          awsAccount: 'aws-account-foo',
          pricingPlan: 'PREMIUM',
          region: 'us-east-1',
        },
        consumer: {
          team: 'team-foo',
          manager: 'Tom Ripen',
          costCenter: 'cs-foo',
        },
      },
    },
    {
      name: 'project-bar',
      namespace: 'namespace-foo',
      repositories: 5,
      labels: {
        provider: {
          awsAccount: 'aws-account-foo',
          pricingPlan: 'PREMIUM',
          region: 'us-east-1',
        },
        consumer: {
          team: 'team-bar',
          manager: 'Rose Dawson',
          costCenter: 'cs-bar',
        },
      },
    },
  ],
};

{
  project:: {
    schema:: projectSchemaFunc,
    example:: projectExample,
  },
  projectCreate:: {
    schema:: projectCreateSchemaFunc,
    example:: projectCreateExample,
  },
  projectList:: {
    schema:: projectListSchemaFunc,
    example:: projectListExample,
  },
}
