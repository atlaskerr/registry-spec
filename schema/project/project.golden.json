{
   "labels": {
      "costCenter": "cs-foo",
      "manager": "Frank Ponnez",
      "team": "team-example"
   },
   "name": "project-example",
   "repositories": 34
}
